src/storm/utility/permutation.cpp

```diff
@@ -0,0 +1,139 @@
+    switch (order) {
+        case OrderKind::Bfs:
+            return "bfs";
+        case OrderKind::Dfs:
+            return "dfs";
+        case OrderKind::ReverseBfs:
+            return "reverse-bfs";
+        case OrderKind::ReverseDfs:
+            return "reverse-dfs";
+        case OrderKind::Random:
+            return "random";
+    }
+    STORM_LOG_ASSERT(false, "unreachable");
+    return "";
+}
+
+OrderKind orderKindFromString(std::string const& order) {
+    for (auto kind : {OrderKind::Bfs, OrderKind::Dfs, OrderKind::ReverseBfs, OrderKind::ReverseDfs, OrderKind::Random}) {
+        if (order == orderKindtoString(kind)) {
+            return kind;
+        }
+    }
+    STORM_LOG_ASSERT(false, "Unknown order kind");
+    return OrderKind::Bfs;
+}
+
+std::vector<std::string> orderKinds() {
+    std::vector<std::string> kinds;
+    for (auto kind : {OrderKind::Bfs, OrderKind::Dfs, OrderKind::ReverseBfs, OrderKind::ReverseDfs, OrderKind::Random}) {
+        kinds.push_back(orderKindtoString(kind));
+    }
+    return kinds;
+}
+
+std::vector<index_type> createRandomPermutation(index_type size) {
+    std::random_device rd;
+    return createRandomPermutation(size, rd());
+}
+
+std::vector<index_type> createRandomPermutation(index_type size, index_type seed) {
+    std::vector<index_type> permutation(size);
+    std::iota(permutation.begin(), permutation.end(), 0);
+    // The random number engine below assumes that it operates on 64-bit unsigned integers.
+    static_assert(std::is_same_v<index_type, uint64_t>, "index_type is assumed to be a 64-bit unsigned integer");
+    std::shuffle(permutation.begin(), permutation.end(), std::mt19937_64(seed));
+    return permutation;
+}
+
+template<typename ValueType>
+std::vector<index_type> createPermutation(OrderKind order, storm::storage::SparseMatrix<ValueType> const& transitionMatrix,
+                                          storm::storage::BitVector const& initialStates) {
+    if (order == OrderKind::Random) {
+        return createRandomPermutation(transitionMatrix.getRowGroupCount());
+    }
+    STORM_LOG_ASSERT((order == OrderKind::Bfs || order == OrderKind::Dfs || order == OrderKind::ReverseBfs || order == OrderKind::ReverseDfs),
+                     "Unknown order kind");
+    STORM_LOG_ASSERT(initialStates.size() == transitionMatrix.getRowGroupCount(), "Unexpected dimensions of initial states and transition matrix.");
+    STORM_LOG_ASSERT(initialStates.size() == transitionMatrix.getColumnCount(), "Unexpected dimensions of initial states and transition matrix.");
+    std::vector<index_type> permutation;
+    permutation.reserve(transitionMatrix.getRowGroupCount());
+
+    std::deque<index_type> stack(initialStates.begin(), initialStates.end());
+    storm::storage::BitVector discoveredStates = initialStates;
+    bool const useFifoStack = order == OrderKind::Bfs || order == OrderKind::ReverseBfs;
+
+    while (!stack.empty()) {
```

Comment:
dont we have graph algorithms which give BFS and DFS for you?

Reply:
True, we have implemented these kind of BFS/DFS traversals at other places (for example [storm/utility/graph.cpp](https://github.com/moves-rwth/storm/blob/master/src/storm/utility/graph.cpp)).
However, I don't think we can easily get the permutation out of that.

We could eventually make something that is more modular and usable in all kind of graph algorithms. But imo this is something for a future refactor of graph.h

Reply:
Ah, I see, I thought you obtain a sorted list of states and then you compute the permuation from that?

Reply:
If I see this correctly, that sorted list of states would already be the inverse of the permutation.